### PR TITLE
Letter template attach pages button and view

### PR DESCRIPTION
**Pull request by CrystalPea**
Adds an "Attach pages" button to the template page. It links to the "Attach pages" page, that contains guidance and a "Choose file" upload button. All hidden behind a feature flag.

Questions: is letter specification right for this use case? Since the attachment doesn't need to have a first page with address block.

### Button:
Screenshot 2023-03-10 at 13 57 07

### New page:
Screenshot 2023-03-10 at 13 30 41

Story card: https://trello.com/c/R5LHg26V/291-add-attach-pages-letter-template-page

### CODE REVIEW

File: app/templates/views/templates/template.html [hunk 1]
```diff
@@ -44,6 +44,21 @@
...
+      "text": "Attach pages",
+      "href": url_for(
+          '.letter_template_attach_pages',
+          service_id=current_service.id,
+          template_id=template.id
+        ),
+      "classes": "govuk-button--secondary govuk-!-margin-right-3"
+      }) }}
+    </div>
+  {% endif %}
```
Comment (samuelhwilliams) [inline app/templates/views/templates/template.html]: This button still looks a little cramped underneath, do we maybe want a bit more padding/margin on the bottom?

File: tests/app/main/views/test_templates.py [hunk 2]
```diff
@@ -776,6 +776,47 @@ def test_view_letter_template_does_not_display_send_button_if_template_over_10_p
...
+    page = client_request.get(
+        "main.view_template",
+        service_id=SERVICE_ONE_ID,
+        template_id=fake_uuid,
+        _test_page_title=False,
+    )
+
+    buttons = [b for b in page.select(".govuk-button--secondary") if normalize_spaces((b).text) == "Attach pages"]
+
+    if template_page_count < 10 and template_type == "letter":
```
Comment (samuelhwilliams) [inline tests/app/main/views/test_templates.py:811]: I'm a bit tempted to say that it might be nice to split these out into separate tests. I don't _hate_ having one conditional check around some assertions but having two makes me think that we're testing quite different things/cases. Having them separate and with clear names around the good/bad thing under test might be better. I don't mind a bit of repetition/duplication in test cases.

File: app/templates/views/templates/attach-pages.html [hunk 3]
```diff
@@ -0,0 +1,51 @@
+{% extends "withnav_template.html" %}
+{% from "components/file-upload.html" import file_upload %}
+{% from "components/page-header.html" import page_header %}
+{% from "govuk_frontend_jinja/components/back-link/macro.html" import govukBackLink %}
+
+{% set page_title = "Attach pages" %}
+
+{% block service_page_title %}
+	{{ page_title }}
```
Comment (samuelhwilliams) [inline app/templates/views/templates/attach-pages.html:9]: omg it's an actual tab character 😱  haven't seen one of those in a while...

Comment (samuelhwilliams) [reply app/templates/views/templates/attach-pages.html:9]: Looks like all of this file is tabs which is pretty unusual for our project.

File: app/templates/views/templates/template.html (hunk 1 above)
Comment (CrystalPea) [reply app/templates/views/templates/template.html]: You are right, the prototype has a bit more padding under the button, I will add some more padding 👍🏼

File: tests/app/main/views/test_templates.py [hunk 4]
```diff
@@ -776,6 +776,64 @@ def test_view_letter_template_does_not_display_send_button_if_template_over_10_p
     assert page.select_one("h1", {"data-error-type": "letter-too-long"})
 
 
+@pytest.mark.parametrize("template_page_count", [6, 10])
+@pytest.mark.parametrize("template_type", ["letter", "email", "sms"])
+def test_view_letter_template_has_attach_pages_button_if_template_below_10_pages_long(
```
Comment (samuelhwilliams) [inline tests/app/main/views/test_templates.py:781]: I think there are some sticky edge cases that you should think about here as epic lead (but potentially could be handled separately). If the template is 9 pages long with the blank personalisation it could hit 10 pages with personalisation injected (esp if some of the personalisation is eg an address that spans multiple lines). How do we want to handle that?

I think the button disappearing altogether is maybe weird - possibly we show show the button but it flags an error saying 'you can't attach pages as the letter would be too long', or the button is greyed out, or something. But disappearing altogether might be odd?

File: app/templates/views/templates/template.html [hunk 5]
```diff
@@ -44,6 +44,21 @@
     {% include 'views/templates/_template.html' %}
   </div>
 
+  {% if template.template_type == "letter" and page_count < 10 and show_attach_pages_button %}
+    <div class="bottom-gutter-1-2">
```
Comment (samuelhwilliams) [inline app/templates/views/templates/template.html:48]: Sorry, I've only just realised - do we still want to use this class? GOV.UK Design System classes I'm more used to using for this sort of thing are `govuk-!-margin-bottom-[1-9]`. @tombye do you have thoughts?

File: tests/app/main/views/test_templates.py (hunk 2 above)
Comment (CrystalPea) [reply tests/app/main/views/test_templates.py:811]: Hmm on one side yes, but I think it's actually easier to look through as a single test. It has so much set up that stays the same regardless of which case we look at. I'd keep it as a single test.

**Comment by samuelhwilliams** [comment]
I think on the whole this looks good 👍  A couple of niggly bits and some sidebar questions, but I think broadly this is in a good place. Once everything's addressed please re-review tag me :)

File: tests/app/main/views/test_templates.py (hunk 4 above)
Comment (CrystalPea) [reply tests/app/main/views/test_templates.py:781]: > If the template is 9 pages long with the blank personalisation it could hit 10 pages with personalisation injected (esp if some of the personalisation is eg an address that spans multiple lines). How do we want to handle that?

That feels a bit out of scope. Also we don't do anything like this when letter template is exactly 10 pages long - we don't warn users it can easily go over based on placeholders.
I think as long as there is a space to add an attachment, we should offer that option, and then if the whole would breach 10 pages upon sending the letter due to placeholder values, we will throw an error then.

> I think the button disappearing altogether is maybe weird - possibly we show show the button but it flags an error saying 'you can't attach pages as the letter would be too long', or the button is greyed out, or something. But disappearing altogether might be odd?

Hmm... We could show a message instead a button, like: "this letter template is at maximum length, no space left for an attachment". If we do that it will make sense to split those test cases you mentioned, indeed. I will ask @karlchillmaid about the right content.

File: app/templates/views/templates/template.html (hunk 5 above)
Comment (CrystalPea) [reply app/templates/views/templates/template.html:48]: I just copied what was used already in that file for other bottom items 😳 . If we change it, I think it would be good to change it in all the files that use it at once, which is out of scope for this PR.

File: tests/app/main/views/test_templates.py (hunk 4 above)
Comment (samuelhwilliams) [reply tests/app/main/views/test_templates.py:781]: Also, one edge case you definitely need to have in mind is what happens if someone edits the template after attaching some pages and that pushes the page over the 10 pages limit.

File: app/templates/views/templates/template.html (hunk 5 above)
Comment (samuelhwilliams) [reply app/templates/views/templates/template.html:48]: I'm not particularly bothered if you leave this, then, but I probably don't agree.

If Tom agrees that govuk-!-margin-x is the more correct/preferred thing, then using that means we aren't making the situation worse, and we're being more clear about what we want to be using going forward. We don't have to fix everything all at once before we can start doing things better. I would prefer to use the 'new' way now, and clean up old stuff later.

I'd be fine with leaving this as-is in this PR if you're saying you're happy to go through and replace the bottom-gutter classes separately "soon", but I don't think that's what we're saying

Comment (tombye) [reply app/templates/views/templates/template.html:48]: It would be great if you could use govuk-!-margin-bottom-3` instead, assuming doing so doesn't break anything. There is [a story to remove these classes](https://www.pivotaltracker.com/story/show/171603272) I haven't ported across from pivotal yet, so it will get done but it would be really useful if any new code we introduce could use the override classes from GOVUK Frontend instead, just so we don't increase the work involved.

It's worth noting that the `bottom-gutter-x-x` classes don't just add margin. They also apply a 'clearfix' to the element (see https://github.com/alphagov/notifications-admin/blob/main/app/assets/stylesheets/_grids.scss#L54). This is an old hack to stop any child elements set to float from poking out of the parent. Because of this, I'd try a straight swap but let me know if it breaks anything as I could help fix it.

File: app/templates/views/templates/attach-pages.html [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+{% block maincolumn_content %}
+<div class="govuk-grid-row">
+    <div class="govuk-grid-column-five-sixths">
+        {{ page_header(page_title) }}
+
+        <p class="govuk-body">Upload a PDF and we’ll print it as part of your letter.</p>
+        <p class="govuk-body">
+            In total, your letter must be 10 pages or less (5 double-sided sheets of paper).
+            You can attach up to 9 pages.
+        </p>
```
Comment (karlchillmaid) [inline app/templates/views/templates/attach-pages.html:27]:
```suggestion
        <p class="govuk-body">
            In total, your letter must be 10 pages or less (5 double-sided sheets of paper).
        </p>
```

File: app/templates/views/templates/attach-pages.html [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        <p class="govuk-body">
+            Attached pages will start on the next available page.
+            If you want to start your attachment on a new sheet of paper, add 2 line breaks
+        </p>
```
Comment (karlchillmaid) [inline app/templates/views/templates/attach-pages.html:31]:
```suggestion
```

File: app/templates/views/templates/attach-pages.html [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        <p class="govuk-body">
+            Your file must meet our <a class="govuk-link govuk-link--no-visited-state" href="{{ url_for('main.guidance_upload_a_letter') }}">letter
+                specification</a>.
+        </p>
+        <p class="govuk-body">Page size and layout: A4 portrait (210 × 297 mm)</p>
+        <p class="govuk-body">Maximum file size: 2 MB</p>
+        <p class="govuk-body">Your letter must be 10 pages or less (5 double-sided sheets of paper).</p>
```
Comment (karlchillmaid) [inline app/templates/views/templates/attach-pages.html:38]:
```suggestion
```

File: tests/app/main/views/test_templates.py (hunk 4 above)
Comment (CrystalPea) [reply tests/app/main/views/test_templates.py:781]: Hmmm... I think the same should happen that we do when they try to edit the template to go over 10 pages by itself. Worth checking if we don't allow them to save such changes, or if we block them when they are sending.

File: app/templates/views/templates/template.html (hunk 5 above)
Comment (CrystalPea) [reply app/templates/views/templates/template.html:48]: Thanks both, I will try swapping it and let you know what I find.

Comment (CrystalPea) [reply app/templates/views/templates/template.html]: All looks good I think.

Screenshot 2023-03-14 at 17 55 04

**Comment by CrystalPea** [comment]
Thanks for content review @karlchillmaid 🙏🏼 

Here is a screenshot of the updated page:

Screenshot 2023-03-14 at 18 12 27

File: tests/app/main/views/test_templates.py (hunk 4 above)
Comment (CrystalPea) [reply tests/app/main/views/test_templates.py:781]: I will add a ticket for this.

Comment (CrystalPea) [reply tests/app/main/views/test_templates.py:781]: Voilla: https://trello.com/c/PqMqV88F/222-for-a-template-with-an-attachment-show-an-error-if-their-total-number-of-pages-is-too-long 😊

**Review by samuelhwilliams — APPROVED** [review]
